### DISCUSSION TITLE
Internationalisation / i18n - testing and guidelines 🌐

### DISCUSSION THREAD

**Discussion by lb-** [wagtail/wagtail#8017] [Development]
In the context of https://github.com/wagtail/wagtail/discussions/7689 & https://github.com/wagtail/wagtail/discussions/7739 we are seeking advice, recommendations and feedback on how to continue to support multiple languages in the Wagtail editor.

## Overview

This **discussion the Wagtail editor and its support for multiple languages, cultures and locales**. Not about the translation of Wagtail pages or content by editors building content within Wagtail for others, e.g. not about Wagtail's ability to [support multi-lingual content](https://docs.wagtail.org/en/stable/advanced_topics/i18n.html).

We know there are many areas for improvement and existing bugs (linked below), however as we embark on some clean up work and rethinking the frontend architecture we want to better understand things.

Such as what problems we have not yet thought about and what are the pain points for users of the Wagtail editor. Additionally, what tooling is out there that we could adopt to make this easier to get right, we know automated tools are not perfect but they can help.

## Areas

**These are just ideas to get us started.**

For each of the areas below, we are seeking feedback, ideas and recommendations. Some of these may end up in our official documentation or code CI. Adding **the items below is not an indicator that it will become a core part of the code** or development approach.

### 1. Guidelines for developers (contributing to Wagtail code)

* Status: **Needs ideas**
* 1.1 Empathy - hard one to quantify or document, not sure on naming, but as we develop we need to remember that our language and our culture is not going to be the same for everyone, try to avoid making assumptions about things like Left to Right, pluralisation, try to avoid colloquialisms in documentation. Remember that when looking at other languages that you do not speak/read, you will not be able to see the things that native speakers see.
* 1.2 Translations
  * all strings must be translated and be easy to translate at the Python & JS level.
  * Consistent approach to JS side translations - see WIP https://github.com/wagtail/wagtail/pull/7953
  * Never do string sorting in JS (e.g. for sortable tables) see [the Turkish problem](https://www.transifex.com/blog/2015/crazy-localization-problems/)
  * Run `makemessages` with `--no-location` flag. This removes the filename and line number comments from the po files. This prevents merge conflicts.
  * Override make messages command to fix POT-Creation-Date PO file header to a default date. This header is changed each time you run makemessages. But is not used (except by humans, but they should use git ;)). Again This prevents merge conflicts.
  * Use `--no-fuzzy-matching` flag. Gettext will mess things up. Humans do a better job at matching.
  * Add a test from django.utils import translation and with translation.override("SOME_LOCALE"): to test for a specific sting in a specific locale.
  * Mark translations with `[ 'Translators:' comments. ](https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#comments-for-translators)` In Python and in templates: `{% comment %}Translators: A time period on a single date{% endcomment %}`
  * Watch out for single item translations clashing with pluralization translations. Eg: `{% blocktranslate count counter=list|length %}Project{% plural %}Projects{% endblocktranslate %}` will clash with: `{% trans 'Project' %}` And also clash with: `{% trans 'Projects' %}`
  * Keep texts as simple as possible. No pages have been created at this location. Why not create one? Doesn't contribute much as it duplicates functionality.
  * Be consistent 'Add a workflow' and 'New workflow task' Pick one and stick with it. 
* 1.3 Casing (upper/sentence/proper-nouns)
  * `verbose_name` and model naming should not be written as a proper noun (e.g. `document`, not `Document`), unless it has a proper noun in it (e.g. `Wagtail documents`). We may refer to models as proper nouns inside documentation, but that is for a different purpose. verbose name should be able to used within a sentence at any point.
  * When needing to output something in sentence case (first letter capitalised), use Django's `capfirst` util or template tag filter, do not use `capitalize` or `upper` as this can cause problems in other languages or even break the meaning of the word as it may intentionally have mixed casing (think of names like McIntosh becoming MCINTOSH). There are exceptions for this such as when needing to compare values, but upper casing should never happen across a whole string for presentation to the user.
  * Similar to the usage of `capitalize` in Python, there should be no need to use this when presenting strings/text to the user from JavaScript (`toUpperCase`, or `toLocaleUpperCase`).
* 1.4 Fonts - System font usage for supporting multiple languages
  * https://github.com/wagtail/wagtail/pull/7972
  * Font tokens (sizes/weights) should consider other languages and minimum font size should not be too small
  * Font Styling - avoid italics?, avoid [uppercase labels](https://github.com/wagtail/wagtail/issues/7624) as uppercase does not mean the same thing in all languages and can case problems with things that should not have capitals in some places.
  * Encoding - should be good but something to double check
* 1.5 Locale
  * Always use full locale, should we check for anywhere that this is not the case?
* 1.6 Dates & locale specific items
  * Date picker and date presentation seems to be a constant source of bugs
  * We should never hard-code date formats
* 1.7 Styles & Design (CSS)
  *  Should we adopt [`quotes` styling](https://www.smashingmagazine.com/2014/06/css-driven-internationalization-in-javascript/#the-lang-pseudo-class) [MDN quotes](https://developer.mozilla.org/en-US/docs/Web/CSS/quotes)
  * Formal support for RTL. No support for [top-to-bottom (TTB) scripts](https://www.w3.org/International/questions/qa-scripts.en) (Simplified & traditional Chinese, Kanji, Hiragana, Katakana, Hangul, Hanja) – as they’re all frequently written LTR as well and a CMS interface can be expected to stay as LTR for those languages.
  * This includes consideration for ions that have directionality (e.g. anything with right or left in the icon name)
  * Design and build frontend for variable text lengths.  For example, when translating English into French or Spanish the translation will generally be about 15 % to 30 %  longer than the original. For German or Dutch it can even be up to 35 % or more. Compound words make this worse. English “Input processing features” may become “Eingabeverarbeitungsfunktionen” in German and "Invoerverwerkingsfuncties" in Dutch.
    * There are also cases where text would be shorter, for example in Arabic scripts. This is straight from https://rtlstyling.com/posts/rtl-styling#3.-differences-in-word-sizes-between-languages. They recommend considering a min-width in some cases to help with this.
  * Based on https://rtlstyling.com/posts/rtl-styling (recommended by @fabienheureux), here are some relevant items
  * We need to be careful with custom letter-spacing, which often [causes issues in Arabic](https://rtlstyling.com/posts/rtl-styling#1.-letter-spacing).
  * Another common issue is [underlines overlapping with diactritics](https://rtlstyling.com/posts/rtl-styling#2.-underlined-links).
  * Suggest enforcing is to never use `float`. As it does not have good cross-browser support for logical values - see https://caniuse.com/mdn-css_properties_float_flow_relative_values.
* 1.8 Images / Icons
  * Never use images that contain text (probably only applicable to the documentation) where the text is meant to be read to gain meaning.
  * Icons as a supporting mechanism but never the only way to communicate something.
  * We’ll need to document the need to horizontally mirror the icons that have directionality (e.g. anything with right or left in the icon name).

### 2. Translation writing

* Status: **Already exploring**
* Note: This section should probably move up into Section 1 - the 'guidelines' and then all subsequent sections are how do we outwork the guidelines.
* 2.1 Documentation for translators
  * We could provide better support for those that do translation contributions https://github.com/wagtail/wagtail/issues/7115
* 2.2 Provide context for translators
  *  https://github.com/wagtail/wagtail/issues/7111 
  * Using the approaches (https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#comments-for-translators & https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#translator-comments-in-templates).
  * Transifex apparently has support for screenshots (see https://docs.transifex.com/api-2-5/screenshots and https://transifex.github.io/openapi/index.html#tag/Context-Screenshots), it would be amazing if we were able to show a screenshot of the original text demonstrating its usage. This would greatly reduce the barrier-to-entry for new translators.
  * Puppeteer for headless browser tests, perhaps it could be set up to automatically capture screenshots of tested pages.
* 2.3 Avoid unnecessary white-space translations
  * Resolve potential existing translation issues https://github.com/wagtail/wagtail/issues/7113

### 3. Code linting

* Status **Already exploring**
* 3.1 [Eslint](https://github.com/wagtail/eslint-config-wagtail)
  * jsx - [enforce no string literals](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-literals.md) - could be a simple rule to ensure we use translations
  * general TS/JS files - enforce no string literals similar to [i18n-next](https://github.com/dgraham/eslint-plugin-i18n-text)
  * Is it possible to enforce that we never concatenate strings
  * Never do alpha sorting clientside (not sure if this is possible to lint) - or provide a translation compatible JS util for this?
* 3.2 [Stylelint](https://github.com/wagtail/stylelint-config-wagtail)
  * text-align/margins that support RTL out of the box
  * enforce correct line-height to show the letters completely and not cut out
  * enforce fonts cannot get too small
  * enforce not adding font-family anywhere except the core usage?
  * do not allow whitespace; nowrap (or maybe just when mixed with overflow hidden)
* 3.4 flake8 / Python linting
  * Any recommendations on ensuring we always use translations (CI should fail for bare strings?)
  * Is it possible to enforce that we never concatenate strings
  * Do not allow synonyms for common things (add not create for example), to attempt to have a consistent vocabulary to build on and translate. (we would want to do the same in JS and HTML)
* 3.5 Django HTML
  * Somehow enforce that we never use `{% blocktrans %}` without trimmed arguments. -> `{% blocktrans trimmed %}   as it will ignore leading and trailing white.   {% endblocktrans %} ` -> can be done with Curlylint.
* 3.6 Django Checks / Runtime analysis
  * It’s a runtime thing, not static analysis, but I believe it’s possible to have a Django check that all form field labels are translate-able. (https://github.com/wagtail/wagtail/issues/7114)

### 4. Unit testing

* Status: **Ideas needed**
* Any recommendations for ways to ensure that our unit tests better cover translations, beyond locale handling, is this something that has value?
* We could add some Python tests for devs to add all translation strings to po files
```
@pytestmark
def test_has_all_migrations():
    output = StringIO()
    management.call_command('makemigrations', verbosity=1, interactive=False, stdout=output)

    output.seek(0)
    output = output.read().strip()
    assert output == 'No changes detected'
```

### 5. Automated tests (aka integration tests, headless browser tests)

* Status: **Critical gap**
* Currently we are using [Puppeteer](https://github.com/puppeteer/puppeteer) (Chrome headless browser)
* Recommendations for using Puppeteer to check multiple languages? Not just that translations worked but that text is not cut off somehow?
* How do we validate that our content (boxes/buttons/layout) wont break when using different languages, can we check for truncated content?
* Should we explore visual regression testing for this? https://testingbot.com/resources/articles/playwright-visual-regression-testing / https://loki.js.org/getting-started.html
* Potentially flag a few languages that should be the go-to for manual testing. Most likely ones that has much longer words than English (Dutch, German, Finnish?), and one(s) that are RTL (Arabic?).

### 6. Manual testing

* Status: **Critical gap**
* Should there be a reference list of translations we recommend be validated against?
* How can this be done easily locally, is just changing the browser local / local user language setting enough to do some rough validation?
* Browserstack makes it easier to switch languages, however this is something we need ideas for.
* What kinds of things should someone look out for if they do not speak another language when doing manual testing on other languages? or is it something that you cannot do as you simply will not know what to look for?
* How can we make it easier to do manual testing, getting to the right audience?
* Potentially flag a few languages that should be the go-to for manual testing. Most likely ones that has much longer words than English (Dutch, German, Finnish?), and one(s) that are RTL (Arabic?).

### 7. Code reviews

* Status: **Ideas needed**
* Not sure how we do this but maybe we can add some reference to guidelines in new PRs
* Maybe we can add a label for 'needs translation review'

### Resources

* https://www.smashingmagazine.com/2012/07/12-commandments-software-localization/
* https://www.w3.org/International/questions/qa-i18n
* https://www.softwaretestinghelp.com/7-basic-tips-for-testing-multi-lingual-web-sites/
* https://www.mediawiki.org/wiki/Internationalization/Styling_and_typography
* https://github.com/wagtail/wagtail/wiki/Managing-Wagtail-translations
* https://www.smashingmagazine.com/2014/12/how-to-conduct-website-localization/
* https://uxdesign.cc/learnings-from-designing-for-multi-language-user-interfaces-573bcb688eee
* https://docs.djangoproject.com/en/4.0/topics/i18n/
* https://docs.djangoproject.com/en/4.0/topics/i18n/translation/
* https://www.browserstack.com/guide/localization-testing-on-websites-and-apps
* https://phrase.com/blog/posts/10-common-mistakes-in-software-localization/
* https://www.transifex.com/blog/2015/crazy-localization-problems/
* https://rtlstyling.com/posts/rtl-styling
* https://chenhuijing.com/blog/css-for-i18n/
* https://docs.transifex.com/
* https://transifex.github.io/openapi/index.html
* https://spectrum.adobe.com/page/bi-directionality/ - example of bi-directional components (RTL)

### Additional discussions or issues

* https://github.com/wagtail/wagtail/discussions/7793
* https://github.com/wagtail/wagtail/issues/7624
* https://github.com/wagtail/wagtail/issues/7117
* https://github.com/wagtail/wagtail/pull/7543
* https://github.com/wagtail/wagtail/issues/7856
* https://github.com/wagtail/wagtail/discussions/7691
* https://github.com/wagtail/wagtail/pull/7953
* https://github.com/wagtail/wagtail/issues/7115
* https://github.com/wagtail/wagtail/issues/7856
* https://github.com/wagtail/wagtail/issues/5504
* https://github.com/wagtail/wagtail/issues/5454
* https://github.com/wagtail/wagtail/issues/4846
* https://github.com/wagtail/wagtail/issues/7111
* https://github.com/wagtail/wagtail/issues/7112
* https://github.com/wagtail/wagtail/issues/7113
* https://github.com/wagtail/wagtail/issues/7114
* https://github.com/wagtail/wagtail/issues/7115
* https://github.com/wagtail/wagtail/issues/7116
* https://github.com/wagtail/wagtail/issues/7117
* https://github.com/wagtail/wagtail/issues/9219
* https://github.com/wagtail/wagtail/issues/9113
* https://github.com/wagtail/wagtail/issues/9193
* https://github.com/wagtail/wagtail/issues/9045

**Comment by olifante** [comment]
Great outline with lots of sensible proposals. Having spent a lot of time translating stuff on Transifex, one thing I found sorely lacking was a lack of context, especially for single-word or short text fragments. Even as someone familiar with Django and Wagtail, a lot of these short texts can be very ambiguous and hard to translate without additional context (is it a noun, a verb, an adjective?).

It would be invaluable for translators if the Part-of-Speech of short text fragments was indicated in comments, using the approaches that @lb- indicated (https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#comments-for-translators & https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#translator-comments-in-templates).

Since Transifex apparently has support for screenshots (see https://docs.transifex.com/api-2-5/screenshots and https://transifex.github.io/openapi/index.html#tag/Context-Screenshots), it would be amazing if we were able to show a screenshot of the original text demonstrating its usage. This would greatly reduce the barrier-to-entry for new translators.

Since we already use Puppeteer for headless browser tests, perhaps it could be set up to automatically capture screenshots of tested pages. @allcaps mentioned on #translators that he has a screen capture project on hold which he might revive.

**Reply by lb-** [reply]
@olifante  - this is great, thank you, I have done my best to incorporate these into the sections above.

**Comment by thibaudcolas** [comment]
This is looking amazingly comprehensive @lb-! Here are my suggestions for amendments & additions:

## RTL

> Formal support, or at least consideration, of RTL & vertical writing.

RTL writing will be formally supported pretty soon. For [top-to-bottom (TTB) scripts](https://www.w3.org/International/questions/qa-scripts.en) (Simplified & traditional Chinese, Kanji, Hiragana, Katakana, Hangul, Hanja) – they’re all frequently written LTR as well. I’ve never seen a site using TTB direction for text for things like forms / tables / advanced interfaces, just text, so I suspect people would expect a CMS interface to stay LTR for those languages.

> Design and build frontend for variable text lengths

All of your examples are about text being longer than english. There are also cases where text would be shorter, for example in Arabic scripts. This is straight from https://rtlstyling.com/posts/rtl-styling#3.-differences-in-word-sizes-between-languages, I dunno how common this is in practice. They recommend considering a `min-width` in some cases to help with this.

> Images / Icons

Either here or alongside RTL info – we’ll need to document the need to horizontally mirror the icons that have directionality (e.g. anything with `right` or `left` in the icon name)

> stylelint

An additional thing I’d suggest enforcing is to never use `float`, which I have recently learned doesn’t have good cross-browser support for logical values (oops). See https://caniuse.com/mdn-css_properties_float_flow_relative_values.

---

In addition to the above, based on https://rtlstyling.com/posts/rtl-styling (recommended by @fabienheureux), I gathered the following tidbits that are relevant for Wagtail:

- We need to be careful with custom `letter-spacing`, which often [causes issues in Arabic](https://rtlstyling.com/posts/rtl-styling#1.-letter-spacing).
- Another common issue is [underlines overlapping with diactritics](https://rtlstyling.com/posts/rtl-styling#2.-underlined-links).

## Other things

> Python linting

It’s a runtime thing, not static analysis, but I believe it’s possible to have a Django check that all form field labels are translate-able. (#7114)

> Django HTML – blocktrans trimmed

I’ll add a linting rule for this to Curlylint.

> Testing (manual & automated)

Re testing, my inclination is to earmark a few languages that should be the go-to for manual testing. Most likely ones that has much longer words than English (Dutch, German, Finnish?), and one(s) that are RTL (Arabic?). And we can do some of the same with automated tests too.

> Resources

I’d add https://rtlstyling.com/posts/rtl-styling on Fabien’s recommendation

> Additional discussions or issues

- #7111
- #7112
- #7113
- #7114
- #7115
- #7116 
- #7117

cc @loicteixeira, would be lovely if you could review this thread.

**Reply by lb-** [reply]
@thibaudcolas thanks, I have done my best to get these into the right sections above.

I think the content should be reworked to section 1 being reference guidelines, section 2 should be part of the guidelines and then 3/4/5/6 are kind of 'how we enact those guidelines' in an automated or manual way.

**Reply by loicteixeira** [reply]
You linked all the issues I had created on the subject, so I don't have much more to add. I simply need to find the time to work on those 😅

Regarding the upload of screenshots for context, last time I checked it was a paid option. I've been thinking of a DjangoDebugToolbar style plugin which would register string sources when rendering a page and it's somehow searchable. But yet again, I need to find some time.

**Comment by hazho** [comment]
just an input for the . .
> All of your examples are about text being longer than english. There are also cases where text would be shorter...
 
in all cases, the most of titles/nav-menu/headers needs to be fixed in min and max range of width, the best approach for this, is to check([with python) the prefered language of the user, and change the font-size based on the language but not the min-width/max-width

**Reply by thibaudcolas** [reply]
@hazho oh, I’ve never seen this done before. Do you have an example of a site that does this, or a source that explains this technique in more details? If you changed the font size of the whole UI based on the language – how would you decide what size to set it to for different languages?

**Reply by hazho** [reply]
for example I have the template that checks the meta agent (if it firefox font size  of :root is 15px but if it is chromium-based is 16px) I used this earlier, due to some problems related layout sizing, because we use (rem and em) unit measurements, but now firefox updated and the issue gone.
thinking of the language, I'll try to prepare a code for that with some examples multi-language-based templates

**Reply by hazho** [reply]
https://gist.github.com/hazho/5fd17fa7a5fa5e5eec1a47817c08d65a
here a simple example, you can test with both  included languages (en, ar) and try to add more languages and see how it goes.
note that there are two options in the <head><style>

**Comment by lb-** [comment]
Nice high level primer on internationalisation and localisation - cannot see anything not already covered but I like the terminology of being 'culturally conscious'.

https://prototypr.io/post/dont-alienate-your-user-a-primer-for-internationalisation-localisation/

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/pull/8226

PR submitted to trim whitespace when translations are used.

**Comment by lb-** [comment]
> Wagtail 4.0 issue that is worth flagging for learnings here.

## Translation for choosers changed from 'choose another' to 'change'.

**Pre 4.0 (English)**

* 'Chooser another' was used and not 'change'

**4.0 (English)**
![187713067-d9a04730-1ba2-4f01-ac51-85537919baa4](https://user-images.githubusercontent.com/1396140/190007760-ee1ae9b1-f998-4d34-98e6-e775b415a6a8.png)
<img width="709" alt="Screen Shot 2022-09-14 at 6 58 58 am" src="https://user-images.githubusercontent.com/1396140/190008011-7dcab160-d5f8-4ee3-a41a-74da5fce9d63.png">

**4.0 (French)**

![189930766-c5552b6d-112c-4505-91f8-2c20526bdd50](https://user-images.githubusercontent.com/1396140/190008738-191198c1-0df8-4bee-9592-f2d0fe80dfd1.png)


**4.0 RC (Italian)**

Note: Now fixed but shows the issue with the underlying word usage.

![185446603-352c8fe0-f4cc-47fa-bbd1-f4da03ef47b5](https://user-images.githubusercontent.com/1396140/190008618-aac112c1-56db-478d-8c52-40ca35ffefa9.png)


## Details

* `change` and `edit` are similar words, even in English, but in other languages these can be translated to basically the same thing.
* The French translation translates both "Edit" and "Change" to "Modifier"
* The Italian translation was initially (during 4.0 RC) using the word "Modifica" for both these labels

## Issues / PRs

* Bug report https://github.com/wagtail/wagtail/issues/9113
* Duplicate bug report https://github.com/wagtail/wagtail/issues/9193
* Issue raised during 4.0 RC for translating these to the same thing https://github.com/wagtail/wagtail/issues/9045
* Fix https://github.com/wagtail/wagtail/pull/9194
* Text change introduced based on UI designs for new page editor work via https://github.com/wagtail/wagtail/pull/8934

## What we can learn from this

* How things will be translated is not always clear, but using similar words to mean the same thing is probably a bad idea.
* Working out when this will be a problem is not always easy - see https://en.wikipedia.org/wiki/Word-sense_disambiguation
* When we see designs, we should be careful to adopt text changes verbatim from designs and understand the why behind the text change or maybe leave text changes for existing elements as is unless the UI is actually changing drastically.

**Reply by thibaudcolas** [reply]
Thank you @lb-, that’s a pretty interesting one to ponder. In this case I think there’s clearly a problem with translations choosing to conflate two concepts while the english source doesn’t, so I’ve created a separate issue to resolve this: #9201. Either we should use the same term everywhere (at the expense of clarity), or we should consistently use different terms in all languages.

**Reply by lb-** [reply]
Thanks Thibaud. I agree that the separate issue makes sense ('change' / 'edit' for the types of permissions) and hopefully we can address this also.

But this comment was more to the original problem created by using 'change' when we originally had 'choose another X' or similar, I think that we can be a bit more careful when adopting a 'simpler' label for something in the future.

**Comment by lb-** [comment]
Added section 1.3 to call out some guidelines for casing words.
---

* 1.3 Casing (upper/sentence/proper-nouns)
  * `verbose_name` and model naming should not be written as a proper noun (e.g. `document`, not `Document`), unless it has a proper noun in it (e.g. `Wagtail documents`). We may refer to models as proper nouns inside documentation, but that is for a different purpose. verbose name should be able to used within a sentence at any point.
  * When needing to output something in sentence case (first letter capitalised), use Django's `capfirst` util or template tag filter, do not use `capitalize` or `upper` as this can cause problems in other languages or even break the meaning of the word as it may intentionally have mixed casing (think of names like McIntosh becoming MCINTOSH). There are exceptions for this such as when needing to compare values, but upper casing should never happen across a whole string for presentation to the user.
  * Similar to the usage of `capitalize` in Python, there should be no need to use this when presenting strings/text to the user from JavaScript (`toUpperCase`, or `toLocaleUpperCase`).

---

Feedback welcome @thibaudcolas  @gasman  @th3hamm0r  - as per discussion on https://github.com/wagtail/wagtail/issues/9219

**Comment by lb-** [comment]
@banjo-tech has submitted a great PR that gives us a starting place for Translator's onboarding https://github.com/wagtail/wagtail/pull/9431 - which resolves most of the initial request for this on https://github.com/wagtail/wagtail/issues/7115 - Thanks !

**Reply by lb-** [reply]
@banjo-tech - have a read through this whole discussion, maybe suggest a small incremental improvement we can make that aligns with these goals. From there we can give feedback and raise an issue for you to work on.

**Comment by lb-** [comment]
@loicteixeira has done an incredible chunk of work to make it easier for translators to translate content for Wagtail. 

Essentially enforcing a consistent way that variables are used (CI will fail if translations are not aligned) and adoption of that convention for all of Wagtail. 

See

- https://github.com/wagtail/wagtail/issues/7111
- https://github.com/wagtail/wagtail/pull/9377
- https://github.com/wagtail/wagtail/commit/5c1c2c8f531d96f4568f6dfa6ce71bc32dd9d16c
- https://docs.wagtail.org/en/latest/contributing/translations.html#marking-strings-for-translation
